action/flow/instance/auditevents.go

```diff
@@ -0,0 +1,224 @@
+package instance
+
+import (
+	"github.com/TIBCOSoftware/flogo-contrib/action/flow/model"
+	"sync"
+)
+
+type Status string
+
+const (
+	Created   = "Created"
+	Completed = "Completed"
+	Cancelled = "Cancelled"
+	Failed    = "Failed"
+	Scheduled = "Scheduled"
+	Skipped   = "Skipped"
+	Started   = "Started"
+	Waiting   = "Waiting"
+	Unknown   = "Created"
+)
+
+type FlowEventListenerFunc func(*FlowEventContext)
+type TaskEventListenerFunc func(*TaskEventContext)
+
+var flowEventListeners []FlowEventListenerFunc
+var taskEventListeners []TaskEventListenerFunc
+
+var lock = &sync.Mutex{}
+
+// FlowEventContext provides access to flow instance execution details
+type FlowEventContext struct {
+	flowInstance *Instance
+}
+
+// Returns flow name
+func (fe *FlowEventContext) Name() string {
+	return fe.flowInstance.Name()
+}
+
+// Returns flow ID
+func (fe *FlowEventContext) ID() string {
+	return fe.flowInstance.ID()
+}
+
+// In case of subflow, returns parent flow name
+func (fe *FlowEventContext) ParentName() string {
+	if fe.flowInstance.master != nil {
+		return fe.flowInstance.master.Name()
+	}
+	return ""
+}
+
+// In case of subflow, returns parent flow ID
+func (fe *FlowEventContext) ParentID() string {
+	if fe.flowInstance.master != nil {
+		return fe.flowInstance.master.ID()
+	}
+	return ""
+}
+
+// Returns current flow status
+func (fe *FlowEventContext) Status() Status {
+	return convertFlowStatus(fe.flowInstance.Status())
+}
+
+//TODO: Should we read once?
+// Returns flow runtime attributes
+func (fe *FlowEventContext) GetWorkingData() map[string]interface{} {
```

Comment:
@debovema : I believe state can be maintained in the listener itself. No need to add something back into the working data.

Reply:
In fact I used Working Data in the first place because I was lacking unique instance IDs for flows and tasks. Now there are available so I can manage OT objects' states elsewhere.

Reply:
Sounds good, so lets get rid of GetWorkingData.

Reply:
@fm-tibco @debovema : I kept it because of iterator use case where one may want get current iteration details.